File: cracking-the-code-interview/chapter16/bfs.js
const bfs = matrix => {
    const visited = [];
    for (let i = 0; i < matrix.length; i++) {
        visited[i] = [];

        for (let j = 0; j < matrix[0].length; j++) {
            visited[i][j] = null;
        }
    }
};

const matrix = [];
for (let i = 0; i < 10; i++) {
    matrix[i] = [];

    for (let j = 0; j < 10; j++) {
        matrix[i][j] = -1;
    }
}

console.log(bfs(matrix));

File: search/graph/bfs.py
adjacency_list = {
    "5": ["3", "7"],
    "3": ["2", "4"],
    "7": ["8"],
    "2": [],
    "4": ["8"],
    "8": []
}


def bfs(start, end):
    if start == end:
        return True

    root = adjacency_list.get(start)
    visited = [start]
    stack = [root]

    while stack:
        neighbors = stack.pop()

        for neighbor in neighbors:
            if neighbor not in visited:
                if neighbor == end:
                    return True, visited
                else:
                    visited.append(neighbor)
                    stack.append(adjacency_list.get(neighbor))

    return False, visited


print(bfs("5", "4"))
